### PR TITLE
Typescript/action sheet

### PR DESCRIPTION
**Pull request by mendyEdri**
## Description
Migrating `ActionSheet` and ActionSheet's demo screen to typescript

## Changelog
Migrating `ActionSheet` and ActionSheet's demo screen to typescript.

### CODE REVIEW

File: src/components/actionSheet/index.tsx [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+        onModalDismissed={onModalDismissed}
+        panDirection={PanningProvider.Directions.DOWN}
+      >
+        {this.renderSheet()}
+      </Dialog>
+    );
+  }
+}
+
+export default asBaseComponent(ActionSheet);
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:249]: This should be:
```suggestion
export default asBaseComponent<ActionSheetProps>(ActionSheet);
```
Once this is done you can run `npm run build:dev` to create the `.d.ts` file and (after you've added the change to the `generatedTypes/index`) you'll have proper autocomplete for it in `ActionSheetScreen`.

File: src/components/actionSheet/index.tsx [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+   */
+  dialogStyle?: StyleProp<ViewStyle>;
+  /**
+   * Add or override style of the options list
+   */
+  optionsStyle?: StyleProp<ViewStyle>;
+  /**
+   * Render custom title
+   */
+  renderTitle?: () => JSX.Element | JSX.Element[];
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:73]: Not sure about this: `| JSX.Element[]`

File: src/components/actionSheet/index.tsx [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  /**
+   * Render custom action
+   * Note: you will need to call onOptionPress so the option's onPress will be called
+   */
+  renderAction?: (
+    option: ButtonProps,
+    index: number,
+    onOptionPress: ActionSheetOnOptionPress
+  ) => JSX.Element | JSX.Element[];
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:82]: Not sure about this: | JSX.Element[]

File: src/components/actionSheet/index.tsx [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+   */
+  testID?: string;
+};
+
+/**
+ * @description: Cross platform Action Sheet, with a support for native iOS solution
+ * @gif: https://media.giphy.com/media/l0HUpXOR6RqB2ct5S/giphy.gif
+ * @example: https://github.com/wix/react-native-ui-lib/blob/master/demo/src/screens/componentScreens/ActionSheetScreen.js
+ */
+class ActionSheet extends Component<ActionSheetProps, {}> {
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:102]:
```suggestion
class ActionSheet extends Component<ActionSheetProps> {
```

File: src/components/actionSheet/index.tsx [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    super(props);
+
+    this.onOptionPress = this.onOptionPress.bind(this);
+    this.renderAction = this.renderAction.bind(this);
+  }
+
+  static defaultProps = {
+    title: undefined,
+    message: undefined,
+    useModal: false
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:115]: Was this intended?

File: src/components/actionSheet/index.tsx [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+      if (showCancelButton) {
+        optionsArray.push({label: 'Cancel'});
+        cancelBtnIndex = optionsArray.length - 1;
+      }
+
+      ActionSheetIOS.showActionSheetWithOptions(
+        {
+          title,
+          message,
+          options: _.flatMap(optionsArray, 'label'),
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:137]: This was `_.map` before, was this intentional?

File: src/components/actionSheet/index.tsx [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+  onOptionPress(optionIndex: number) {
+    _.invoke(this.props, `options[${optionIndex}].onPress`);
+    _.invoke(this.props, 'onDismiss');
+  }
+
+  handleRenderIcon = (option: ButtonProps) => {
+    // @ts-ignore
+    let source = option.icon;
+    if (!source) {
+      source = _.isFunction(option.iconSource) ? option.iconSource() : option.iconSource as ImageProps['source'];
```
Comment (M-i-k-e-l) [inline src/components/actionSheet/index.tsx:155]: This was not in the JS version, does it fix a previous JS bug or a typescript error?

Comment (mendyEdri) [reply src/components/actionSheet/index.tsx:155]: That was a bug. the `butttonProps` contains `iconSource` and not `icon` - as it is now.
We decided to be compatible with both of them for now.

File: src/components/actionSheet/index.tsx (hunk 6 above)
Comment (mendyEdri) [reply src/components/actionSheet/index.tsx:137]: Yes, `_.map` can return nulls elements inside an array, which we don't support, nor with the type we declared.
(`_.flatMap` will return array with defined values or an empty array)

File: src/components/actionSheet/index.js [hunk 8]
```diff
@@ -1,252 +0,0 @@
-import _ from 'lodash';
```
[the remark this answers is not in the document]
Comment (mendyEdri) [reply src/components/actionSheet/index.js:1]: I did as described in the guidelines..
also running `git log --follow -p -- src/components/actionSheet/index.tsx` do log all file history

File: src/components/actionSheet/index.tsx [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+import {ActionSheetIOS, StyleSheet, StyleProp, ViewStyle, ImageProps, ImageSourcePropType} from 'react-native';
+import {Constants} from '../../helpers';
+import {Colors} from '../../style';
+import {asBaseComponent} from '../../commons/new';
+import Dialog, {DialogProps} from '../dialog';
+import View from '../view';
+import Text from '../text';
+import {ButtonProps} from '../button';
+import Image from '../image';
+import ListItem from '../listItem';
```
[the remark this answers is not in the document]
Comment (mendyEdri) [reply src/components/actionSheet/index.tsx:12]: What do we usually do about it?

File: src/components/actionSheet/index.tsx (hunk 5 above)
Comment (mendyEdri) [reply src/components/actionSheet/index.tsx:115]: removed

File: src/components/actionSheet/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply src/components/actionSheet/index.tsx:12]: I usually start with the inner components, not sure about this one, are you going to continue with this task?

File: src/components/actionSheet/index.tsx (hunk 6 above)
Comment (M-i-k-e-l) [reply src/components/actionSheet/index.tsx:131]: On second thought, why would anyone send an `option` with `label=null`?

File: generatedTypes/index.d.ts [hunk 10]
```diff
@@ -70,7 +70,6 @@ export {default as Carousel, CarouselProps} from './components/carousel';
 
 /* All components with manual typings */
 export {
-  ActionSheet,
```
[the remark this answers is not in the document]
Comment (M-i-k-e-l) [reply generatedTypes/index.d.ts:73]: Missed this?

File: src/components/actionSheet/index.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (mendyEdri) [reply src/components/actionSheet/index.tsx:13]: Consulted with Ethan, ignoring for now.

File: src/components/actionSheet/index.tsx (hunk 6 above)
Comment (mendyEdri) [reply src/components/actionSheet/index.tsx]: Fixed, thank you

**Review by M-i-k-e-l — APPROVED** [review]

**Pull request merged**